app/oauth_test.go

```diff
@@ -143,3 +147,355 @@ func TestOAuthDeleteApp(t *testing.T) {
 		t.Fatal("should not get session from cache or db")
 	}
 }
+
+func TestAuthorizeOAuthUser(t *testing.T) {
+	setup := func(enable, tokenEndpoint, userEndpoint bool, serverURL string) *TestHelper {
+		th := Setup()
+
+		th.App.UpdateConfig(func(cfg *model.Config) {
+			cfg.GitLabSettings.Enable = enable
+
+			if tokenEndpoint {
+				cfg.GitLabSettings.TokenEndpoint = serverURL + "/token"
+			} else {
+				cfg.GitLabSettings.TokenEndpoint = ""
+			}
+
+			if userEndpoint {
+				cfg.GitLabSettings.UserApiEndpoint = serverURL + "/user"
+			} else {
+				cfg.GitLabSettings.UserApiEndpoint = ""
+			}
+		})
+
+		return th
+	}
+
+	makeState := func(token *model.Token) string {
+		return base64.StdEncoding.EncodeToString([]byte(model.MapToJson(map[string]string{
+			"token": token.Token,
+		})))
+	}
+
+	makeToken := func(th *TestHelper, cookie string) *model.Token {
+		token, _ := th.App.CreateOAuthStateToken(generateOAuthStateTokenExtra("", "", cookie))
+		return token
+	}
+
+	makeRequest := func(cookie string) *http.Request {
```

Comment:
I think we should check `err` here.

```suggestion
  makeRequest := func(t *testing.T, cookie string) *http.Request {
		request, err := http.NewRequest(http.MethodGet, "https://mattermost.example.com", nil)
		require.Nil(t, err)
```

Reply:
Yeah, I can add that. While this is just test code and it should never fail because the URL is a hardcoded valid URL (as far as I know, `http.NewRequest` only returns an error for bad URLs), it's probably a good habit to get into